src/Builder/Encoder/OperatorEncoder.php

```diff
@@ -0,0 +1,161 @@
+            case Encode::Single:
+                return $this->encodeAsSingle($value);
+
+            case Encode::Array:
+                return $this->encodeAsArray($value);
+
+            case Encode::Object:
+                return $this->encodeAsObject($value);
+
+            case Encode::DollarObject:
+                return $this->encodeAsDollarObject($value);
+
+            case Encode::Group:
+                assert($value instanceof GroupStage);
+
+                return $this->encodeAsGroup($value);
+        }
+
+        throw new LogicException(sprintf('Class "%s" does not have a valid ENCODE constant.', $value::class));
+    }
+
+    /**
+     * Encode the value as an array of properties, in the order they are defined in the class.
+     */
+    private function encodeAsArray(OperatorInterface $value): stdClass
+    {
+        $result = [];
+        /** @var mixed $val */
+        foreach (get_object_vars($value) as $val) {
+            // Skip optional arguments.
+            // $slice operator has the optional <position> argument in the middle of the array
+            if ($val === Optional::Undefined) {
+                continue;
+            }
+
+            $result[] = $this->recursiveEncode($val);
+        }
+
+        return $this->wrap($value, $result);
+    }
+
+    /**
+     * $group stage have a specific encoding because the _id argument is required and others are variadic
+     */
+    private function encodeAsGroup(GroupStage $value): stdClass
+    {
+        $result = new stdClass();
+        $result->_id = $this->recursiveEncode($value->_id);
+
+        foreach (get_object_vars($value->field) as $key => $val) {
+            $result->{$key} = $this->recursiveEncode($val);
+        }
+
+        return $this->wrap($value, $result);
+    }
+
+    private function encodeAsObject(OperatorInterface $value): stdClass
+    {
+        $result = new stdClass();
+        foreach (get_object_vars($value) as $key => $val) {
+            // Skip optional arguments. If they have a default value, it is resolved by the server.
+            if ($val === Optional::Undefined) {
+                continue;
+            }
+
+            $result->{$key} = $this->recursiveEncode($val);
+        }
+
+        return $this->wrap($value, $result);
+    }
+
+    private function encodeAsDollarObject(OperatorInterface $value): stdClass
+    {
+        $result = new stdClass();
+        foreach (get_object_vars($value) as $key => $val) {
+            // Skip optional arguments. If they have a default value, it is resolved by the server.
+            if ($val === Optional::Undefined) {
+                continue;
+            }
+
+            $val = $this->recursiveEncode($val);
+
+            if ($key === 'geometry') {
+                if (is_object($val) && property_exists($val, '$geometry')) {
+                    $result->{'$geometry'} = $val->{'$geometry'};
+                } elseif (is_array($val) && array_key_exists('$geometry', $val)) {
```

Comment:
Seeing these side by side, I just realized that `property_exists()` and `array_key_exists()` use inconsistent argument ordering. Oh, PHP.